mrmustard/training/trainer.py

```diff
@@ -0,0 +1,422 @@
+            example, when device-making is contained in `cost_fn`). Defaults to None.
+        metric_fns (Union[Sequence[callable], Mapping[callable], callable]): Optional collection of functions that takes the
+            output of `device_factory` after optimization and returns arbitrary evaluation/information.
+        return_kwargs (bool): Whether to include input config `kwargs` in the output dict. Defualts to True.
+        skip_opt (bool): Whether to skip the optimization and directly calculate cost.
+        tag (str): Optional label of the training task associated with the `kwargs` to be included in the output dict.
+        kwargs: Dict containing all arguments to any of the functions below:
+            - `cost_fn`: exluding the output of `device_factory`.
+            - `device_factory`: e.g. `x`, `r`, `theta`, etc.
+            - `Optimizer`: e.g. `euclidean_lr`.
+            - `Optimizer.minimize`: excluding `cost_fn` and `by_optimizing`, e.g. `max_steps`.
+
+    Returns:
+        dict: A result dict summarizing the optimized circuit, cost, metrics and/or input configs.
+
+    """
+
+    setting_updates, kwargs = update_pop(mm.settings, **kwargs)
+
+    input_kwargs = kwargs.copy() if return_kwargs else {}
+
+    device, kwargs = (
+        curry_pop(device_factory, **kwargs) if callable(device_factory) else ([], kwargs)
+    )
+    device = [device] if not isinstance(device, (Sequence, Mapping)) else device
+
+    cost_fn, kwargs, optimized = _apply_partial_cost(device, cost_fn, **kwargs)
+
+    opt = None
+    if optimized and not skip_opt:
+        opt, kwargs = curry_pop(Optimizer, **kwargs)
+        _, kwargs = curry_pop(
+            opt.minimize, **{"cost_fn": cost_fn, "by_optimizing": optimized}, **kwargs
+        )
+
+    if kwargs:
+        warnings.warn(f"Unused kwargs: {kwargs}")
+
+    final_cost = cost_fn()
+
+    results = {
+        "cost": np.array(final_cost).item(),
+        "device": device,
+        "optimizer": opt,
+    }
+
+    if callable(metric_fns):
+        results["metrics"] = metric_fns(*device)
+    elif isinstance(metric_fns, Sequence):
+        results["metrics"] = [f(*device) for f in metric_fns if callable(f)]
+    elif isinstance(metric_fns, Mapping):
+        results = {
+            **results,
+            **{k: f(*device) for k, f in metric_fns.items() if callable(f)},
+        }
+
+    return {
+        **({"tag": tag} if tag is not None else {}),
+        **results,
+        **input_kwargs,
+        **setting_updates,
+    }
+
+
+def _iter_futures(futures):
+    """Make ray futures iterable for easy passing to a progress bar.
+    Hacky: https://github.com/ray-project/ray/issues/5554
+    """
+    import ray  # pylint: disable=import-outside-toplevel
+
+    while futures:
+        done, futures = ray.wait(futures)
+        yield ray.get(done[0])
+
+
+def map_trainer(trainer=train_device, tasks=1, pbar=True, unblock=False, num_cpus=None, **kwargs):
```

Comment:
Can you add a short explanation that in practice the most common use case is to ignore the `trainer` keyword (as it defaults to `train_device`), pbar, unblock etc and just concentrate on `tasks` and `**kwargs`?